### PR TITLE
Fix the parsing issue in large post request for transport-simple-http

### PR DESCRIPTION
**Pull request by jasonjoo2010**
### Describe what this PR does / why we need it
Please refer to #1253 

Improve unit test for HttpEventTask.

### Does this pull request fix one issue?
Yes.

### Describe how you did it
Change single read into iteration.

### Describe how to verify it
Do post request larger than 10kB.

### Special notes for reviews
Nope

### CODE REVIEW

**Comment by jasonjoo2010** [comment]
It's a quite old bug and simple pr so i think we should get it done soon.

Refer to #1253

**Comment by sczyh30** [comment]
I'll test the PR this weekend :)

**Comment by sczyh30** [comment]
I've tested with the code and when I was publishing the rule via the dashboard, an error occurred:

```
2020-03-09 12:32:02.377 WARNING [SimpleHttpCommandCenter] CommandCenter error
java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.read1(BufferedReader.java:212)
	at java.io.BufferedReader.read(BufferedReader.java:286)
	at java.io.Reader.read(Reader.java:140)
	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.consumePostBody(HttpEventTask.java:215)
	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.processPostRequest(HttpEventTask.java:147)
	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.run(HttpEventTask.java:83)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

The error occurred in the second time of reading:

https://github.com/alibaba/Sentinel/blob/0c9150ae06b65eb42fdda5351dca51c189343d8b/sentinel-transport/sentinel-transport-simple-http/src/main/java/com/alibaba/csp/sentinel/transport/command/http/HttpEventTask.java#L210-L217

It seems that the SocketInputStream will wait until timeout, rather than for EOF. Maybe we need to check it via `in.ready()` or other approaches? @jasonjoo2010

**Comment by jasonjoo2010** [comment]
> I've tested with the code and when I was publishing the rule via the dashboard, an error occurred:
> 
> ```
> 2020-03-09 12:32:02.377 WARNING [SimpleHttpCommandCenter] CommandCenter error
> java.net.SocketTimeoutException: Read timed out
> 	at java.net.SocketInputStream.socketRead0(Native Method)
> 	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
> 	at java.net.SocketInputStream.read(SocketInputStream.java:171)
> 	at java.net.SocketInputStream.read(SocketInputStream.java:141)
> 	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
> 	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
> 	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
> 	at java.io.InputStreamReader.read(InputStreamReader.java:184)
> 	at java.io.BufferedReader.fill(BufferedReader.java:161)
> 	at java.io.BufferedReader.read1(BufferedReader.java:212)
> 	at java.io.BufferedReader.read(BufferedReader.java:286)
> 	at java.io.Reader.read(Reader.java:140)
> 	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.consumePostBody(HttpEventTask.java:215)
> 	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.processPostRequest(HttpEventTask.java:147)
> 	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.run(HttpEventTask.java:83)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> The error occurred in the second time of reading:
> 
> https://github.com/alibaba/Sentinel/blob/0c9150ae06b65eb42fdda5351dca51c189343d8b/sentinel-transport/sentinel-transport-simple-http/src/main/java/com/alibaba/csp/sentinel/transport/command/http/HttpEventTask.java#L210-L217
> 
> It seems that the SocketInputStream will wait until timeout, rather than for EOF. Maybe we need to check it via `in.ready()` or other approaches? @jasonjoo2010

I have restructured it dealing with the `Content-Length` header. Now it should be solved but i only pass it under unit tests.

I will make actually traffic test tomorrow and update the result. Surely if you still have the codes of scene you can simply retry.

**Comment by jasonjoo2010** [comment]
> I've tested with the code and when I was publishing the rule via the dashboard, an error occurred:
> 
> ```
> 2020-03-09 12:32:02.377 WARNING [SimpleHttpCommandCenter] CommandCenter error
> java.net.SocketTimeoutException: Read timed out
> 	at java.net.SocketInputStream.socketRead0(Native Method)
> 	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
> 	at java.net.SocketInputStream.read(SocketInputStream.java:171)
> 	at java.net.SocketInputStream.read(SocketInputStream.java:141)
> 	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
> 	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
> 	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
> 	at java.io.InputStreamReader.read(InputStreamReader.java:184)
> 	at java.io.BufferedReader.fill(BufferedReader.java:161)
> 	at java.io.BufferedReader.read1(BufferedReader.java:212)
> 	at java.io.BufferedReader.read(BufferedReader.java:286)
> 	at java.io.Reader.read(Reader.java:140)
> 	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.consumePostBody(HttpEventTask.java:215)
> 	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.processPostRequest(HttpEventTask.java:147)
> 	at com.alibaba.csp.sentinel.transport.command.http.HttpEventTask.run(HttpEventTask.java:83)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> The error occurred in the second time of reading:
> 
> https://github.com/alibaba/Sentinel/blob/0c9150ae06b65eb42fdda5351dca51c189343d8b/sentinel-transport/sentinel-transport-simple-http/src/main/java/com/alibaba/csp/sentinel/transport/command/http/HttpEventTask.java#L210-L217
> 
> It seems that the SocketInputStream will wait until timeout, rather than for EOF. Maybe we need to check it via `in.ready()` or other approaches? @jasonjoo2010

Tests have been passed and detail cases are listed below.
And more codes are restructured: 
* Reader is replaced with plain InputStream to support Content-Type
* readLine() under InputStream
* StatusCode becomes standalone enum to make response clear

## Cases
Cases are run under `sentinel-demo-command-handler`

### Normal Get Request
```
GET /echo?name=test HTTP/1.1
Host: demo.com
Accept: */*
Accept-Language: en-us
Accept-Encoding: plain/text

```
=> 
```
HTTP/1.0 200 OK
Content-Length: 11
Connection: close

Hello: test
```

### Normal Post Request
```
POST /echo?name=fake HTTP/1.1
Host: demo.com
Accept: */*
Accept-Language: en-us
Accept-Encoding: plain/text
Content-Type: application/x-www-form-urlencoded
Content-Length: 9

name=test
```
=>
```
HTTP/1.0 200 OK
Content-Length: 11
Connection: close

Hello: test
```

### Smaller Content-Length
```
POST /echo?name=fake HTTP/1.1
Host: demo.com
Accept: */*
Accept-Language: en-us
Accept-Encoding: plain/text
Content-Type: application/x-www-form-urlencoded
Content-Length: 8

name=test
```
=>
```
HTTP/1.0 200 OK
Content-Length: 10
Connection: close

Hello: tes
```

### Incomplete Request
```
POST /echo?name=fake HTTP/1.1
Host: demo.com
Accept: */*
Accept-Language: en-us
Accept-Encoding: plain/text
Content-Type: application/x-www-form-urlencoded
Content-Length: 100

name=test
```

That depends. If client close the stream server will just accept and parse it normally but if client flush the buffer and wait without closing stream it will lead to a `Read Timeout` error and response as HTTP_500.

### Malformed Request
```
POST /echo?name=fake HTTP/1.1
Host: demo.com
Accept: */*
Accept-Language: en-us
Accept-Encoding: plain/text
Content-Type: application/x-www-form-urlencoded
Content-Length: 20

name=test
s
d
o
fihasdoihfoiasdhfiohasdiof
`
if the body is not encoded correctly server will still accept it(there are illegal newlines in this example)
=>
```
HTTP/1.0 200 OK
Content-Length: 22
Connection: close

Hello: test
s
d
o
fiha
```

File: sentinel-transport/sentinel-transport-simple-http/src/main/java/com/alibaba/csp/sentinel/transport/command/http/HttpEventTask.java [hunk 1]
```diff
@@ -190,56 +303,31 @@ private void closeResource(Closeable closeable) {
...
-    /**
-     * Write `200 OK` HTTP response status line and flush.
-     */
-    private void writeOkStatusLine(/*@NonNull*/ final PrintWriter out) {
-        out.print("HTTP/1.1 200 OK\r\n"
-            + "Connection: close\r\n\r\n");
+    
+    private void writeResponse(PrintWriter out, StatusCode statusCode, String message) {
+        out.print("HTTP/1.0 " + statusCode.toString() + "\r\n"
+                + "Content-Length: " + (message == null ? 0 : message.length()) + "\r\n"
```
Comment (sczyh30) [inline sentinel-transport/sentinel-transport-simple-http/src/main/java/com/alibaba/csp/sentinel/transport/command/http/HttpEventTask.java]: The content length here is wrong when the message contains non-ASCII characters (e.g. Chinese characters). Maybe we should use `message.getBytes().length` instead?

Comment (jasonjoo2010) [reply sentinel-transport/sentinel-transport-simple-http/src/main/java/com/alibaba/csp/sentinel/transport/command/http/HttpEventTask.java]: Oh you are right i forget to correct here.

**Comment by sczyh30** [comment]
Nice work. Thanks!

**Pull request merged**